### ISSUE TITLE
crash when clearing barrel recipe with creative items (?)

### ISSUE THREAD

**Issue by smcpeak** [Herddex/recursive-resource-calculator#1]
Using the attached save file, which requires [Creative Items](https://mods.factorio.com/mod/CreativeItems), Recursive Resource Calculator will crash if I:

* Load the save (which is sandbox mode, no player character).
* Press Alt-C to open the calculator.
* Make a new sheet and switch to it.
* Choose blue science pack and set 100/m rate.
* Right-click to clear the recipe for water (unbarrelling).
* Right-click to clear the recipe for crude oil (also unbarrelling).

I tried reproducing this with a new sandbox game, both with and without Creative Items, but was unable to, so I don't know if Creative Items is needed.

Mod version: Recursive Resource Calculator 1.0.2
Factorio version: 1.1.37 on Windows

Screenshot:

![crash-when-deselecting-oil-recipe](https://user-images.githubusercontent.com/2909643/141455317-61cba243-aecf-4d3e-bb16-dfc466289554.png)

Save file:

[creative.zip](https://github.com/Herddex/recursive-resource-calculator/files/7526975/creative.zip)

**Comment by Herddex** [comment]
Thank you for the detailed report. I will look into it.

**Comment by Herddex** [comment]
I've identified the issue. The table that holds the information for the Totals section is updated in such a way that if an item/fluid's production rate gets set to 0 (or very, very close to 0), it gets deleted from the table (this is done so that previously existing items from sheets don't just leave "ghost" zeros everywhere in the totals section). However, the same thing is not done for the sheets (as evidenced by the presence of the 0/s empty barrels), simply because the sheets get rebuilt anyway on a new calculation and ghosts wouldn't be an issue. Except that they actually are:

The crash was caused by the subtraction of 0 (which exists in the sheet's version of production rates) from nothing (the totals' version of production rates).

So I think the best solution would be to disallow zeros from the sheet's production rates as well. Hopefully this won't cause any problematic side effects...
